zebra-state/src/service/non_finalized_state.rs

```diff
@@ -565,4 +629,130 @@ impl NonFinalizedState {
+
+            match self.chain_count().cmp(&prev_length_bars) {
+                Greater => self
+                    .chain_fork_length_bars
+                    .resize_with(self.chain_count(), howudoin::new),
+                Less => {
+                    let redundant_bars = self.chain_fork_length_bars.split_off(prev_length_bars);
+                    for bar in redundant_bars {
+                        bar.close();
+                    }
+                }
+                Equal => {}
+            }
+
+            // It doesn't matter what chain the bar was previously used for,
+            // because we update everything based on the latest chain in that position.
+            for (chain_length_bar, chain) in
+                std::iter::zip(self.chain_fork_length_bars.iter(), self.chain_iter())
+            {
+                let fork_height = chain
+                    .last_fork_height
+                    .unwrap_or_else(|| chain.non_finalized_tip_height())
+                    .0;
+
+                // We need to initialize and set all the values of the bar here, because:
+                // - the bar might have been newly created, or
+                // - the chain this bar was previously assigned to might have changed position.
+                chain_length_bar
+                    .label(format!("Fork {fork_height}"))
+                    .set_pos(u64::try_from(chain.len()).expect("fits in u64"))
+                    .set_len(u64::from(
+                        zebra_chain::transparent::MIN_TRANSPARENT_COINBASE_MATURITY,
+                    ));
+
+                // display work as bits
+                let mut desc = format!(
+                    "Work {:.1} bits",
+                    chain.partial_cumulative_work.difficulty_bits_for_display(),
+                );
+
+                if let Some(recent_fork_height) = chain.recent_fork_height() {
+                    let recent_fork_length = chain
+                        .recent_fork_length()
+                        .expect("just checked recent fork height");
+
+                    desc.push_str(&format!(
+                        " at {recent_fork_height:?} + {recent_fork_length} blocks"
+                    ));
+                }
+
+                chain_length_bar.desc(desc);
+            }
+        }
+    }
+
+    /// Stop tracking metrics for this non-finalized state and all its chains.
+    pub fn disable_metrics(&mut self) {
+        #[cfg(feature = "getblocktemplate-rpcs")]
+        {
+            self.should_count_metrics = false;
+        }
+
+        #[cfg(feature = "progress-bar")]
+        {
+            if let Some(chain_count_bar) = self.chain_count_bar.as_ref() {
+                chain_count_bar.close();
+            }
+
+            self.chain_count_bar = None;
+
+            for chain_length_bar in self.chain_fork_length_bars.iter() {
+                chain_length_bar.close();
+            }
+
+            self.chain_fork_length_bars = Vec::new();
```

Comment:
Nitpick: use `take()` and `drain()` here.

```suggestion
            let count_bar = self.chain_count_bar.take().into_iter();
            let fork_bars = self.chain_fork_length_bars.drain(..);
            count_bar.chain(fork_bars).for_each(howudoin::Tx::close);
```